README

Geo::Point
===========

Geo::Point tries to abstract coordinate systems.  It does not try
to solve all the problems itself, but will call-out for helper
modules when computation has to be done.

The module is havily resting on Math::Polygon.

***** This module is only in an BETA development state... it is
***** currently only used with wgs84, so all transformations have
***** not been tested!!!

Improvements and additions please to
       Mark Overmeer, <geopoint@overmeer.net>


ChangeLog
Revision history for Perl extension Geo::Point

version 0.99:

	Changes:
	- use Gis::Distance to replace deprecated Geo::Distance [Aran Deltac]

	Fixes:
	- distance computation reversed coordinates [Aran Deltac]

version 0.98: Fri Feb  1 23:31:14 CET 2019

	Fixes:
	- fix metadata [Mohammad S Anwar]
	- doc typos, rt.cpan.org#126960 [Xavier Guimard]
	- bbox of Geo::Point [Chad Windnagle]

version 0.97: Tue Jan 23 20:43:58 CET 2018

	Improvements:
	- converted to GIT
	- published via GitHUB
	- remove dependency on Test::POD
	- Geo::Shape::new() and Geo::Point::new() as instance method

version 0.96: Tue Jan 21 14:34:23 CET 2014

	Improvements:
	- more work on docs.

version 0.95: Mon Jan 20 16:54:57 CET 2014

	Fixes:
	- typos.  rt.cpan.org#82902 [Xavier Guimard]
	- Geo::Distance integration
	  rt.cpan.org#92242 [Niko Tyni], http://bugs.debian.org/735744

	Improvements:
	- change documentation style.

version 0.94: Fri Dec 21 12:19:47 CET 2012

	Fixes:
	- fix ::Surface::bbox()

version 0.93: Wed May 19 23:31:05 CEST 2010

	Fixes:
	- ::Surface::geo_outer() and geo_inner() renamed to ::geoOuter()
	  and ::geoInner(), to be consequent with naming convensions.
	  Older names still supported but not documented anymore.

	- Geo::Proj::zoneForUTM() used 'zone' as proj4 option, where
	  it should be '+zone'. [Michael R. Davis]

	- Geo::Proj:zoneForUTM() miscalculated zones with negative
	  longitudes. [Michael R. Davis]

version 0.92: Thu Dec 24 12:08:04 CET 2009

	Fixes:
	- synopsis mistakes in Geo::Proj, spotted by
	  rt.cpan.org#47953 [Don Reid]

version 0.91: Mon May 25 14:31:47 CEST 2009

	Fixes:
	- the Geo::Surface was very seriously broken.

version 0.90: Tue Jan 27 16:45:11 CET 2009

	Fixes:

	- do not use /bin/pwd in t/pod.t

	- detected projection of ::Surface not registered.

version 0.10: Wed Jan 21 10:04:19 CET 2009

	Fixes:

	- projection lost during line creation from geopoints.

	- default projection must be expressed as label, not object.

	- default projection for surface is that of first component.

	- many fixes to Surface initiation.

version 0.09: Fri Jan  2 20:08:35 CET 2009

	Fixes:

	- creation of Geo::Line objects based on Geo::Point objects.
	  rt.cpan.org#42076 [Jonathan Yu]

	- fix call to bestUTMprojection for 'utm' default nick-name

	- repare UTM zone calculation.

	- repare reprojecting of lines.

	Improvement:

	- croak when object is passed as proj parameter: must be a
	  Geo::Proj label.

version 0.08: Mon Mar 31 08:58:30 CEST 2008

	- very stupid bugs when 0 was used as lat or long.  Hit by
	  [Paul Tomblin]

version 0.07: Tue Feb 12 10:00:05 CET 2008

	- use Geo::Distance for Geo::Shape::distance [Mark Lawrence]

	- use Math::Trig for Geo::Shape::distance [Mark Lawrence]

	- added dependency for Math::Trig

	- add '0' before 1 digit minutes and seconds in deg2dm() and
	  deg2dms()

	- GeoPoint is stable for quite some time, but the man-page
	  still contained ALPHA and BETA release warnings.

	- added toString() as prefered over string().

version 0.06: Mon Jun 11 16:55:01 CEST 2007

	- added t/pod.t

version 0.05: Fri Mar  9 14:18:25 CET 2007

	- Converted to use oodist, removing mkdist, mkdoc, version.

	- removed superfluous html/index.html

	- email address now geo@, webpage /geo/

	- all error messages should start with lower-case

version 0.04: Thu Nov 30 15:08:19 CET 2006

	- Removed some (harmless) debugging traps

	- Geo::Point::normalize()

version 0.03: Mon Oct  9 20:04:16 CEST 2006

	- Do not keep Geo::Proj4 object within Geo::Shape, so Geo
	  objects can be used in a database.

	- Newest OODoc produces better man-pages.

version 0.002: Mon Sep 25 16:01:18 CEST 2006

	- fix of Geo::Proj::default projection(): should always return a
	  nickname, not an object.

	- overloading with enabled fallback, which fixes some test
	  problems.  by [Chia-liang Kao].

	- overload cmp in Geo::Proj

version 0.001: Mon Dec 12 12:04:21 CET 2005

	Initial release


Makefile.PL
use ExtUtils::MakeMaker;

require 5.008;

WriteMakefile
  ( NAME        => 'Geo::Point'
  , VERSION     => '0.99'
  , PREREQ_PM   =>
     { Test::More    => 0.47
     , GIS::Distance => 0.14
     , Geo::Proj4    => 1.01
     , Math::Polygon => 1.01
     , Math::Trig    => 1.00
     }
  , AUTHOR      => 'Mark Overmeer'
  , ABSTRACT    => 'Geographical structures'
  , LICENSE     => 'perl_5'

  , META_MERGE =>
      { 'meta-spec' => { version => 2 }
      , resources  =>
          { repository =>
              { type => 'git'
              , url  => 'https://github.com/markov2/perl5-Geo-Point.git'
              , web  => 'https://github.com/markov2/perl5-Geo-Point'
              }
          , homepage => 'http://perl.overmeer.net/CPAN/'
          , license  => [ 'http://dev.perl.org/licenses/' ]
          }
      }
  );

#### the next lines are added for OODoc, which generates the
#### distribution.
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/geo-point/raw
DISTDIR         = ../public_html/geo-point/source
LICENSE         = perl

# for POD
FIRST_YEAR      = 2005
EMAIL           = markov@cpan.org
WEBSITE         = http://perl.overmeer.net/CPAN/

# for HTML
EXTENDS		= ../Geo-Proj4:../Geo-WKT:../Geo-GML:../Geo-ISO19139:../../perl/GeoEOP:../XML-Compile:../XML-Compile-Cache:../Geo-Format-Envisat:../Geo-Format-Landsat:../../perl/DBD-mysql-Geom:../Geo-WKT:../Math-Polygon
HTML_OUTPUT     = ../public_html/geo/html
HTML_DOCROOT    = /geo/html
HTML_PACKAGE    = ../public_html/geo-point/htmlpkg

__POSTAMBLE



lib/Geo/Proj.pm
# This code is part of distribution Geo-Point.  Meta-POD processed with
# OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package Geo::Proj;

use strict;
use warnings;

use Geo::Proj4   ();
use Carp         qw/croak/;

=chapter NAME

Geo::Proj - Handling projections

=chapter SYNOPSIS

 use Geo::Proj;

 my $wgs84 = Geo::Proj->new   # predefined if import()
  ( nick  => 'wgs84'
  , proj4 => '+proj=latlong +datum=WGS84 +ellps=WGS84'
  );

 my $clrk = Geo::Proj->new
  ( nick  => 'clark66'
  , proj4 => [proj => "merc", ellps => "clrk66", lon_0 => -96]
  );

 my $point_wgs84= Geo::Point->latlong(56.12, 4.40, 'wgs84');
 my $point_wgs84= Geo::Point->latlong(56.12, 4.40, $wgs84);

 my $point_clrk = $point_wgs84->in($clrk);
 my $point_clrk = Geo::Proj->to($wgs84, $clrk, $point_wgs84);
 my $point_clrk = Geo::Proj->to($wgs84, 'clark66', $point_wgs84);

=chapter DESCRIPTION
A point on Earth's surface can be represented in many different coordinate
systems.  The M<Geo::Proj4> module wraps the popular Open Source C<libproj>
library to convert between those coordinate systems; a very complex job.

Within a program, however, you like some extra abstraction from that
library: to be able to simply label a point to its system, and then
forget about all transformations which may be necessary.  The label (or
C<nick>) hides all complicated parameters for the actual projection .

WARNING 1: this class will collect all nicks, which means that calling
M<new()> with the same label twice will have the second ignored.

WARNING 2: the wgs84 nickname is predefined, but only if this module is
'used' with import.  So if you decide to use 'require' to dynamically load
this module, then don't forget to call 'import()' yourself, or define the
wgs84 projection yourself.

=chapter OVERLOADING

=overload '""' (stringification)
Returns the nick-name for this projection.
=cut

use overload '""'     => sub { shift->nick }
           , fallback => 1;

=chapter METHODS

=cut

sub import()
{
  Geo::Proj->new
   ( nick  => 'wgs84'
   , proj4 => '+proj=latlong +datum=WGS84 +ellps=WGS84'
   );
}

#-----------------------
=section Constructors

=c_method new [$nick], %options
Create a new object.

=requires nick       LABEL
The abbrevated name for this projection.

=requires proj4      OBJECT|ARRAY|STRING
The ARRAY or STRING will by used to create a M<Geo::Proj4> object
by calling M<Geo::Proj4::new()>.  You may also specify such an
prepared OBJECT.

=option   srid       INTEGER
=default  srid       C<undef>
SRID stands for "Spatial Reference System ID", which is just an index
in a table of spatial descriptions as used by SQL. Only INTEGER values
larger than 0 are permitted.

=option   name       STRING
=default  name       <from proj4>

=cut

my %projections;
my $defproj;

sub new(@)
{   my ($class, %args) = @_;
    my $proj   = $projections{$args{nick} || 'dead'};
    return $proj if defined $proj;

    my $self   = (bless {}, $class)->init(\%args);
    $projections{$self->nick} = $self;
    $defproj ||= $self;
    $self;
}

sub init($)
{   my ($self, $args) = @_;

    my $nick = $self->{GP_nick} = $args->{nick}
        or croak "ERROR: nick required";

    $self->{GP_srid} = $args->{srid};

    my $proj4 = $args->{proj4}
        or croak "ERROR: proj4 parameter required";

    if(ref $proj4 eq 'ARRAY')
    {   $proj4   = Geo::Proj4->new(@$proj4);
        croak "ERROR: cannot create proj4: ".Geo::Proj4->error
            unless $proj4;
    }
    elsif(!ref $proj4)
    {   $proj4   = Geo::Proj4->new($proj4);
        croak "ERROR: cannot create proj4: ".Geo::Proj4->error
            unless $proj4;
    }
    $self->{GP_proj4} = $proj4;
    $self->{GP_name}  = $args->{name};
    $self;
}

#-----------------------
=section Attributes

=method nick
Simple abbreviating of the projection.
=cut

sub nick() {shift->{GP_nick}}

=method name
The full, official name of the projection
=cut

sub name()
{   my $self = shift;
    my $name = $self->{GP_name};
    return $name if defined $name;

    my $proj = $self->proj4;
    my $abbrev = $proj->projection
       or return $self->{nick};

    my $def    = $proj->type($abbrev);
    $def->{description};
}

=ci_method proj4 [ <$nick|$proj4> ]
Returns the projection library handle (a M<Geo::Proj4>) to be used by this
component.  As class method, the $nick is specified for a lookup.  In case
a $proj4 is specified, that is returned.

=examples
 my $wgs84 = Geo::Proj->new(nick => 'wgs84', ...);
 my $wgs84_proj4 = Geo::Proj->proj4('wgs84');
 my $wgs84_proj4 = Geo::Proj->proj4($wgs84);
 my $wgs84_proj4 = $wgs84->proj4;
=cut

sub proj4(;$)
{   my $thing = shift;
    return $thing->{GP_proj4} unless @_;

    my $proj  = $thing->projection(shift) or return undef;
    $proj->proj4;
}

=method srid
The "Spatial Reference System ID" if known.
=cut

sub srid() {shift->{GP_srid}}

#-----------------------
=section Projecting

=c_method projection <$nick|$proj>
Returns the M<Geo::Proj> object, defined with $nick.  In case such an
object is passed in as $proj, it is returned unaffected.  This method is
used where in other methods NICKS or $proj can be used as arguments.

=examples
 my $wgs84 = Geo::Proj->projection('wgs84');
 my $again = Geo::Proj->projection($wgs84);
=cut

sub projection($)
{   my $which = $_[1];
    UNIVERSAL::isa($which, __PACKAGE__) ? $which : $projections{$which};
}

=c_method defaultProjection [ <$nick|$proj> ]
The $nick must be defined with M<new()>.  Returned is the Geo::Proj object
for the default projection.  The default is the first name created,
which probably is 'wgs84' (when import() had a chance)

=cut

sub defaultProjection(;$)
{   my $thing = shift;
    if(@_)
    {   my $proj = shift;
        $defproj = ref $proj ? $proj : $thing->projection($proj);
    }
    $defproj;
}

=c_method listProjections
Returns a sorted lost of projection nicks.
=cut

sub listProjections() { sort keys %projections }

=c_method dumpProjections [$fh]
Print details about the defined projections to the $fh, which
defaults to the selected.  Especially useful for debugging.
=cut

sub dumpProjections(;$)
{   my $class = shift;
    my $fh    = shift || select;

    my $default = $class->defaultProjection;
    my $defnick = defined $default ? $default->nick : '';

    foreach my $nick ($class->listProjections)
    {   my $proj = $class->projection($nick);
        my $name = $proj->name;
        my $norm = $proj->proj4->normalized;
        $fh->print("$nick: $name".($defnick eq $nick ? ' (default)':'')."\n");
        $fh->print("    $norm\n") if $norm ne $name;
    }
}

=ci_method to [<$proj|$nick>], <$proj|$nick>, $point|ARRAY-of-$points
Expects an Geo::Proj to project the $point or $points to.  The work
is done by M<Geo::Proj4::transform()>.  As class method, you have to
specify two nicks or projections.

Be warned that this M<to()> method expects POINTs which are
B<not> M<Geo::Point> objects, but which themselves are an ARRAY
containing X,Y and optionally a Z coordinate.

=examples
 my $p2 = $wgs84->to('utm31-wgs84', $p1);
 my $p2 = $wgs84->to($utm, $p1);
 my $p2 = Geo::Proj->to('wgs84', 'utm31-wgs84', $p1);
=cut

sub to($@)
{   my $thing   = shift;
    my $myproj4 = ref $thing ? $thing->proj4 : __PACKAGE__->proj4(shift);
    my $toproj4 = __PACKAGE__->proj4(shift);
    $myproj4->transform($toproj4, shift);
}

#-----------------------
=section UTM
=cut

# These methods may have been implemented in Geo::Point, however may get
# supported by any external library later.  Knowledge about projections
# is as much as possible concentrated here.

=ci_method zoneForUTM $point
Provided some point, figure-out which zone is most optimal for representing
the point.  In LIST context, zone number, zone letter, and meridian are
returned as separate scalars.  In LIST context, the zone number and letter
are returned as one..

This code is stolen from L<Geo::Coordinates::UTM>, because that module
immediately starts to do computations with this knowledge, which is not
wanted here.  Probably a lot of zones are missing.

=cut

sub zoneForUTM($)
{   my ($thing, $point) = @_;
    my ($long, $lat) = $point->longlat;

    my $zone
     = ($lat >= 56 && $lat < 64)
     ? ( $long <  3   ? undef
       : $long < 12   ? 32
       :                undef
       )
     : ($lat >= 72 && $lat < 84)
     ? ( $long <  0   ? undef
       : $long <  9   ? 31
       : $long < 21   ? 33
       : $long < 33   ? 35
       : $long < 42   ? 37
       :                undef
       )
     : undef;

    my $meridian = int($long/6)*6 + ($long < 0 ? -3 : +3);
    $zone      ||= int(($meridian+180)/6) +1;
 
    my $letter
     = ($lat < -80 || $lat > 84) ? ''
     : ('C'..'H', 'J'..'N', 'P'..'X', 'X')[ ($lat+80)/8 ];

      wantarray     ? ($zone, $letter, $meridian)
    : defined $zone ? "$zone$letter"
    : undef;
}

=ci_method bestUTMprojection $point, [$proj|$nick]
Returns the best UTM projection for some $point.  As class method, you
specify the nickname or the object for the point.

=example
 my $point = Geo::Point->longlat(2.234, 52.12);
 my $proj  = Geo::Proj->bestUTMprojection($point);
 print $proj->nick;    # for instance utm31-wgs84

=cut

sub bestUTMprojection($;$)
{   my ($thing, $point) = (shift, shift);
    my $proj  = @_ ? shift : $point->proj;

    my ($zone, $letter, $meridian) = $thing->zoneForUTM($point);
    $thing->UTMprojection($proj, $zone);
}


=c_method UTMprojection <$datum|$proj|undef>, $zone
The $proj is a M<Geo::Proj> which is used to collect the datum
information from if no $datum was specified explicitly.  It may also be
a string which is the name of a datum, as known by proj4.  Undef will
be replaced by the default projection.

=example
 my $proj = Geo::Proj->UTMprojection('WGS84', 31) or die;
 print $proj->nick;    # for instance utm31-wgs84
=cut

sub UTMprojection($$)
{   my ($class, $base, $zone) = @_;

    $base   ||= $class->defaultProjection;
    my $datum = UNIVERSAL::isa($base, __PACKAGE__) ? $base->proj4->datum :$base;
    $datum  ||= 'wgs84';

    my $label = "utm$zone-\L$datum\E";
    my $proj  = "+proj=utm +zone=$zone +datum=\U$datum\E"
              . " +ellps=\U$datum\E +units=m +no_defs";

    Geo::Proj->new(nick => $label, proj4 => $proj);
}

1;


lib/Geo/Point.pm
# This code is part of distribution Geo-Point.  Meta-POD processed with
# OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package Geo::Point;
use base 'Geo::Shape';

use strict;
use warnings;

use Geo::Proj;
use Math::Trig                  qw/deg2rad/;
use GIS::Distance::Constants    qw/$KILOMETER_RHO/;
use Carp                        qw/confess croak/;

=chapter NAME

Geo::Point - a point on the globe

=chapter SYNOPSIS

 use Geo::Point;

 my $p = Geo::Point->latlong(1,2);
 my $p = Geo::Point->longlat(2,1);

 my $w = Geo::Proj->new(wgs84 => ...);
 my $p = Geo::Point->latlong(1,2, 'wgs84');

 my ($lat, $long) = $p->latlong;
 my ($x, $y) = $p->xy;
 my ($x, $y) = $p->in('utm31-wgs84');

 my $p = Geo::Point->xy(1,2);

=chapter DESCRIPTION
One location on the globe, in any coordinate system.  This package tries
to hide the maths and the coordinate system in which the point is
represented.

One of the most confusing things when handling geometrical data, is
that sometimes latlong, sometimes xy are used: horizontal and vertical
organization reversed.  This package tries to hide this from your
program by providing abstract accessors M<latlong()>, M<longlat()>,
M<xy()>, and M<yx()>.

=chapter METHODS

=section Constructors

=ci_method new %options

=option  latitude  COORDINATE
=default latitude  undef

=option  lat       COORDINATE
=default lat       undef

=option  longitude COORDINATE
=default longitude undef

=option  long      COORDINATE
=default long      undef

=option  x         COORDINATE
=default x         undef

=option  y         COORDINATE
=default y         undef

=cut

sub init($)
{   my ($self, $args) = @_;

    $self->SUPER::init($args);
    $self->{GP_x} = defined $args->{x}    ? $args->{x}
                  : defined $args->{long} ? $args->{long}
                  :                         $args->{longitude};
    $self->{GP_y} = defined $args->{y}    ? $args->{y}
                  : defined $args->{lat}  ? $args->{lat}
                  :                         $args->{latitude};
    $self;
}

=ci_method latlong [ $lat,$long,[$proj] ] | [$proj]
When called as class method, you create a new point.  Provide a LATitude
and LONGitude. The optional PROJection tells in which coordinate system.

As instance method, the latitude and longitude are reported.  You
can ask it to be translated into the $proj coordinate system first.

When $proj is undefined, none is presumed. The project must be specified
as string, which referse to a projection defined by M<Geo::Proj>.
See also M<longlat()>, M<xy()>, and M<yx()>.

=example latlong as class method
 my $wgs84 = Geo::Proj->new(wgs84 => ...);
 my $gp    = Geo::Point->latlong(52.3213, 5.53, 'wgs84');

=example latlong as instance method
 my ($lat, $long) = $gp->latlong('wgs84');

=cut

sub latlong(@)
{   my $thing = shift;

    if(ref $thing)   # instance method
    {   return ($thing->{GP_y}, $thing->{GP_x}) unless @_ > 2;

        my $proj = pop @_;
        return $thing->in($proj)->latlong;
    }

    # class method
    $thing->new(lat => shift, long => shift, proj => shift);
}

=ci_method longlat [ $long,$lat,[$proj] ] | [$proj]
Like M<latlong()>, but with the coordinates reversed.  Some applications
prefer this.
=cut

sub longlat(@)
{   my $thing = shift;

    if(ref $thing)   # instance method
    {   return ($thing->{GP_x}, $thing->{GP_y}) unless @_ > 2;
        my $proj = pop @_;
        return $thing->in($proj)->longlat;
    }

    # class method
    $thing->new(long => shift, lat => shift, proj => shift);
}

=ci_method xy [$x, $y, [$proj] ] | [$proj]
Like M<longlat()> but now for carthesian projections.  Usually, the coordinate
order is reversed.  See also M<yx()>.
=cut

sub xy(@)
{   my $thing = shift;

    if(ref $thing)   # instance method
    {   return ($thing->{GP_x}, $thing->{GP_y}) unless @_ > 2;

        my $proj = pop @_;
        return $thing->in($proj)->xy;
    }

    # class method
    $thing->new(x => shift, y => shift, proj => shift);
}

=ci_method yx [$y, $x, [$proj] ] | [$proj]
Like M<latlong()> but now for carthesian projections.  Usually, the
coordinate order is reversed.  See also M<xy()>.
=cut

sub yx(@)
{   my $thing = shift;

    if(ref $thing)   # instance method
    {   return ($thing->{GP_y}, $thing->{GP_x}) unless @_ > 2;

        my $proj = pop @_;
        return $thing->in($proj)->yx;
    }

    # class method
    $thing->new(y => shift, x => shift, proj => shift);
}

=c_method fromString $string, [$projection]
Create a new point from a $string.  The coordinates can be separated by
a comma (preferably), or blanks.  When the coordinates end on NSEW, the
order does not matter, otherwise lat-long or xy order is presumed.

This routine is very smart.  It understands:

  PROJLABEL VALUE VALUE
  PROJLABEL: VALUE VALUE
  PROJLABEL, VALUE, VALUE
  PROJLABEL: VALUE, VALUE
  VALUE VALUE
  VALUE, VALUE
  utm: ZONE, VALUE, VALUE   # also without commas and ':'
  utm: VALUE, VALUE, ZONE   # also without commas and ':'
  utm: VALUE, VALUE         # also without commas and ':'
  ZONE, VALUE, VALUE        # also without commas and ':'
  VALUE, VALUE, ZONE        # also without commas and ':'

The VALUE must be suitable for projection.  If only two values are
provided, a C<d>, single or double quote, or trailing/leading C<e>, C<w>,
C<n>, C<s> (either lower or upper-case) will force a latlong projection.
Those coordinates must follow the rules of M<dms2deg()>.

=examples point from string

 my $x = 'utm 31n 12311.123 34242.12'; # utm zone 31N
 my $x = '12311.123 34242.12 31';      # utm zone 31
 my $x = '123.123E 12.34';             # wgs84  latlong
 my $x = 'clrk66 123.123 12.34';       # clrk66 latlong
 my $x = '12d34'123.1W 11.1123';       # wgs84  longlat

 my $p = Geo::Point->fromString($x);

 # When parsing user applications, you probably want:
 my $p = eval { Geo::Point->fromString($x) };
 warn $@ if $@;

=error UTM requires 3 values: easting, northing, and zone

=error illegal UTM zone in $string
A UTM zone can be detected at the beginning or at the end of the
input.  It contains a number (from 1 up to 60) and an optional
latitude indication (C up to X, except I and O).

=error undefined projection $proj for $string
The projection you used (or is set as default) is not defined.  See
M<Geo::Proj::new()> about how to defined them.

=error too few values in $string (got @parts)
Most projection require two parameters, but utm requires three (with zone).

=error too many values in $string (got @parts)
Most projection require two parameters, but utm requires three (with zone).

=error dms latitude coordinate not understood: $string
See M<dms2deg()> for permitted formats.

=error dms longitude coordinate not understood: $string
See M<dms2deg()> for permitted formats.

=error illegal character in x coordinate $x
=error illegal character in y coordinate $y

=cut

sub fromString($;$)
{   my ($class, $string, $nick) = @_;

    defined $string or return;
    $string =~ s/^\s+//;
    $string =~ s/\s+$//;
    return () unless length $string;

    # line starts with project label
    $nick   = $1 if $string =~ s/^(\w+)\s*\:\s*//;

    # The line is either split by comma's or by blanks.
    my @parts
      = $string =~ m/\,/
      ? (split /\s*\,\s*/, $string)
      : (split /\s+/, $string);

    # Now, the first word may be a projection.  That is: any non-coordinate,
    # anything which starts with more than one letter.
    if($parts[0] =~ m/^[a-z_]{2}/i)
    {   $nick = shift @parts;          # overrules default
    }

    my $proj;
    if(!defined $nick)
    {   $proj = Geo::Proj->defaultProjection;
        $nick = $proj->nick;
    }
    elsif($nick eq 'utm')
    {   die "ERROR: UTM requires 3 values: easting, northing, and zone\n"
           unless @parts==3;

        my $zone;
        if($parts[0] =~ m/^\d\d?[C-HJ-NP-X]?$/i )
        {   $zone = shift @parts;
        }
        elsif($parts[2] =~ m/^\d\d?[C-HJ-NP-X]?$/i )
        {   $zone = pop @parts;
        }

        if(!defined $zone || $zone==0 || $zone > 60)
        {   die "ERROR: illegal UTM zone in $string";
        }

        $proj = Geo::Proj->UTMprojection(undef, $zone);
        $nick = $proj->nick;
    }
    else
    {   $proj = Geo::Proj->projection($nick)
            or croak "ERROR: undefined projection $nick";
    }

    croak "ERROR: too few values in '$string' (got ".@parts.", expect 2)\n"
       if @parts < 2;

    croak "ERROR: too many values in '$string' (got ".@parts.", expect 2)\n"
       if @parts > 2;

    if($proj->proj4->isLatlong)
    {   my ($lats, $longs)
         = (  $parts[0] =~ m/[ewEW]$/ || $parts[1] =~ m/[nsNS]$/
           || $parts[0] =~ m/^[ewEW]/ || $parts[1] =~ m/^[nsNS]/
           )
         ? reverse(@parts) : @parts;

        my $lat  = $class->dms2deg($lats);
        defined $lat
            or die "ERROR: dms latitude coordinate not understood: $lats\n";

        my $long = $class->dms2deg($longs);
        defined $long
           or die "ERROR: dms longitude coordinate not understood: $longs\n";

        return $class->new(lat => $lat, long => $long, proj => $nick);
    }
    else # type eq xy
    {   my ($x, $y) = @parts;
        die "ERROR: illegal character in x coordinate $x"
            unless $x =~ m/^\d+(?:\.\d+)$/;

        die "ERROR: illegal character in y coordinate $y"
            unless $y =~ m/^\d+(?:\.\d+)$/;

        return $class->new(x => $x, y => $y, proj => $nick);
    }

    ();
}

#----------------
=section Accessors
The accessors only work correctly when you are sure that the point is
in the right coordinate systems.

=method latitude
=method lat
=method longitude
=method long
=method x
=method y
=cut

sub longitude() {shift->{GP_x}}
sub long()      {shift->{GP_x}}
sub latitude()  {shift->{GP_y}}
sub lat()       {shift->{GP_y}}

sub x()         {shift->{GP_x}}
sub y()         {shift->{GP_y}}

#----------------
=section Projections

=cut

sub in($)
{   my ($self, $newproj) = @_;

    # Dirty hacks violate OO, to improve the speed.
    return $self if $newproj eq $self->{G_proj};

    my ($n, $p) = $self->projectOn($newproj, [$self->{GP_x}, $self->{GP_y}]);
    $p ? ref($self)->new(x => $p->[0], y => $p->[1], proj => $n) : $self;
}

=method normalize
Be sure the that coordinates are between -180/180 longitude, -90/90
lattitude.  No changes for non-latlong projections.
=cut

sub normalize()
{   my $self = shift;
    my $p    = Geo::Proj->projection($self->proj);
    $p && $p->proj4->isLatlong or return $self;
    my ($x, $y) = @$self{'GP_x','GP_y'};
    $x += 360 while $x < -180;
    $x -= 360 while $x >  180;
    $y += 180 while $y <  -90;
    $y -= 180 while $y >   90;
    @$self{'GP_x','GP_y'} = ($x, $y);
    $self;
}

#----------------
=section Geometry

=method bbox
The bounding box of a point contains twice itself.
=cut

sub bbox() { @{(shift)}{qw/GP_x GP_y GP_x GP_y/} }

=method area
Always returns zero.
=cut

sub area() { 0 }

=method perimeter
Always returns zero.
=cut

sub perimeter() { 0 }

=method distancePointPoint $gisdist, $units, $point
Compute the distance between the current point and some other $point in
$units.  The $gisdist object will do the calculations.  See M<distance()>.
=cut

# When two points are within one UTM zone, this could be done much
# easier...

sub distancePointPoint($$$)
{   my ($self, $gisdist, $units, $other) = @_;

    my $here  = $self->in('wgs84');
    my $there = $other->in('wgs84');
    my $distance = $gisdist->distance($here->latlong, $there->latlong);

    if ($units eq 'degrees') {
        return( ($distance->km() / $KILOMETER_RHO) * 360 );
    }
    elsif ($units eq 'radians') {
        return deg2rad( ($distance->km() / $KILOMETER_RHO) * 360 );
    }

    return $distance->$units();
}

=method sameAs $other, $tolerance
=error can only compare a point to another Geo::Point
=cut

sub sameAs($$)
{   my ($self, $other, $e) = (shift, shift);

    croak "ERROR: can only compare a point to another Geo::Point"
        unless $other->isa('Geo::Point');

    # may be latlong or xy, doesn't matter: $e is corrected for that
    my($x1, $y1) = $self->xy;
    my($x2, $y2) = $other->xy;
    abs($x1-$x2) < $e && abs($y1-$y2) < $e;
}

=method inBBox $object
Returns a true value if this point is inside the bounding box of
the specified $object.  The borders of the bbox are included.  This is
relatively fast to check, even for complex objects.  When the projections
differ, the point is translated into the $object's coordinate system,
because that one must stay square.

=cut

sub inBBox($)
{   my ($self, $other) = @_;
    my ($x, $y) = $self->in($other->proj)->xy;
    my ($xmin, $ymin, $xmax, $ymax) = $other->bbox;
    $xmin <= $x && $x <= $xmax && $ymin <= $y && $y <= $ymax
}

#----------------
=section Display

=method coordsUsualOrder
Returns the coordinates in the order which is usual for the projection
used.
=cut

sub coordsUsualOrder()
{   my $self = shift;
    my $p    = Geo::Proj->projection($self->proj);
    $p && $p->proj4->isLatlong ? $self->latlong : $self->xy;
}

=method coords
Returns the coordinates in their usual order, formatted as string
with a joining blank;
=cut

sub coords()
{  my ($a, $b) = shift->coordsUsualOrder;
   defined $a && defined $b or return '(none)';

   sprintf "%.4f %.4f", $a, $b;
}

=method toString [$projection]
Returns a string representation of the point, which is also used for
stringification.  The default projection is the one of the point.
=examples
 print "Point: ",$gp->toString, "\n";
 print "Point: $gp\n";   # same

 print "Point: ",$gp->toString('clrk66'), "\n";
=cut

sub toString(;$)
{   my ($self, $proj) = @_;
    my $point;

    if(defined $proj)
    {   $point = $self->in($proj);
    }
    else
    {   $proj  = $self->proj;
        $point = $self;
    }

    "point[$proj](" .$point->coords.')';
}
*string = \&toString;

=method dms [$projection]
Show the point as DMS value-pair.  You must be sure that the coordinate
is a projection for which is it useful to represent the values in DMS.
In SCALAR context, one string is returned.  In LIST context, the values
are returned separately in latlong order.

Be warned, that the returned string may contain single and double quote
characters, which may confuse HTML (see M<dmsHTML()>).

=cut

sub dms(;$)
{   my ($self, $proj) = @_;
    my ($long, $lat)  = $proj ? $self->in($proj)->longlat : $self->longlat;

    my $dmslat  = $self->deg2dms($lat,  'N', 'S');
    my $dmslong = $self->deg2dms($long, 'E', 'W');
    wantarray ? ($dmslat, $dmslong) : "$dmslat, $dmslong";
}

=method dm [$projection]
Like M<dms()>, but doesn't show seconds.
=cut

sub dm(;$)
{   my ($self, $proj) = @_;
    my ($long, $lat)  = $proj ? $self->in($proj)->longlat : $self->longlat;

    my $dmlat  = $self->deg2dm($lat,  'N', 'S');
    my $dmlong = $self->deg2dm($long, 'E', 'W');
    wantarray ? ($dmlat, $dmlong) : "$dmlat, $dmlong";
}

=method dmsHTML [$projection]
Like M<dms()>, but all character which are troublesome for HTML are
translated into character codes.
=cut

sub dmsHTML(;$)
{   my ($self, $proj) = @_;
    my @both = $self->dms($proj);
    foreach (@both)
    {   s/"/\&quot;/g;
        # The following two translations are nice, but IE does not handle
        # them correctly when uses as values in form fields.
        # s/d/\&deg;/g;
        # s/ /\&nbsp;\&nbsp;/g;
    }
    wantarray ? @both : "$both[0], $both[1]";
}

=method dmHTML [$projection]
Like M<dmsHTML()>, but does not show seconds.
=cut

sub dmHTML(;$)
{   my ($self, $proj) = @_;
    my @both = $self->dm($proj);
    foreach (@both)
    {   s/"/\&quot;/g;
        # See dmsHTML above
        # s/d/\&deg;/g;
        # s/ /\&nbsp;\&nbsp;/g;
    }
    wantarray ? @both : "$both[0], $both[1]";
}

=method moveWest
Move a point from the eastern calculations into the western calculations,
resulting in a value below -180.  This is useful when this point is part
of a larger construct, like the corners of a satellite image, which are
both sides of the -180 meridian.

=example moving West
 my $point = Geo::Point->latlong(24, 179);
 $point->moveWest;
 print $point->long;   # -181;
=cut

sub moveWest()
{   my $self = shift;
    $self->{GP_x} -= 360 if $self->{GP_x} > 0;
}


1;
